runtime/compiler/optimizer/SequentialStoreSimplifier.cpp

```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+         }
+      else
+         {
+         /* aloadNode for all loaded bytes must be the exact same node. */
+         if (aloadNode != getALoadReferenceForSeqLoad(processedByteNodes[i]))
+            {
+            if (trace) traceMsg(comp, "Sequential Load Simplification aload mismatch. rootNode: %p, aloadNode(0): %p, aloadNode(%d): %p\n",
+                                rootNode, aloadNode, i, getALoadReferenceForSeqLoad(processedByteNodes[i]));
+            return currentTreeTop;
+            }
+
+         /* basePointerNode for all loaded bytes must be the exact same node (or all NULL). */
+         if (basePointerNode != getBasePointerReferenceForSeqLoad(processedByteNodes[i]))
+            {
+            if (trace) traceMsg(comp, "Sequential Load Simplification basePointerNode mismatch. rootNode: %p, basePointerNode(0): %p, basePointerNode(%d): %p\n",
+                                rootNode, basePointerNode, i, getBasePointerReferenceForSeqLoad(processedByteNodes[i]));
+            return currentTreeTop;
+            }
+         }
+
+      byteConversionNode = getByteConversionNodeForSeqLoad(processedByteNodes[i]);
+      if (trace) traceMsg(comp, "Sequential Load Simplification Candidate - rootNode: %p, byteConversionNode: %p\n", rootNode, byteConversionNode);
+      multValue = getMultValueForSeqLoad(processedByteNodes[i]);
+      if (trace) traceMsg(comp, "Sequential Load Simplification Candidate - rootNode: %p, multValue: %ld\n", rootNode, multValue);
+
+      if (multValue > (1L << ((byteCount - 1) * 8)))
+         {
+         if (trace) traceMsg(comp, "multValue is too high for the given byteCount. rootNode: %p, byteCount: %d, multValue: %ld\n", rootNode, byteCount, multValue);
+         return currentTreeTop;
+         }
+
+      if ((1L << ((byteCount - 1) * 8)) == multValue)
+         {
+         /* Check if most significant byte is signed extended or not. */
+         signExtendResult = checkForSeqLoadSignExtendedByte(processedByteNodes[i]);
+         if (trace) traceMsg(comp, "Sequential Load Simplification Candidate - rootNode: %p, signExtendResult: %d\n", rootNode, signExtendResult);
+         }
+      else
+         {
+         /* Other bytes can not be sign extended. */
+         if (checkForSeqLoadSignExtendedByte(processedByteNodes[i]))
+            {
+            if (trace) traceMsg(comp, "Bytes other than the most significant byte should not be sign extended. rootNode: %p, processedByteNodes[%d]: %p\n", rootNode, i, processedByteNodes[i]);
+            return currentTreeTop;
+            }
+         }
+
+      /*
+       * byteConversionNodes and byteOffsets are sorted such that the least significant byte load is in position 0.
+       * Position 1 has the next most significant load and so on.
+       * multValue is used to determine the significance of the byte load.
+       * If two byte loads have the same multValue, the opt bails out.
+       */
+      for (int j = 0; j < byteCount; j++)
```

Comment:
Is it possible to replace this loop with mapping from multValue to j and check that j is with the right range and that there no duplicates? Then the following loop would not be needed either

Reply:
Yes. I can change this.